packages/dashboard-message-bus/lib/DashboardMessageBus.ts

```diff
@@ -0,0 +1,184 @@
+import WebSocket from "isomorphic-ws";
+import delay from "delay";
+import { EventEmitter } from "events";
+import {
+  base64ToJson,
+  broadcastAndAwaitFirst,
+  broadcastAndDisregard,
+  createMessage,
+  jsonToBase64,
+  startWebSocketServer
+} from "./utils";
+import { promisify } from "util";
+
+interface UnfulfilledRequest {
+  socket: WebSocket;
+  data: WebSocket.Data;
+}
+
+export class DashboardMessageBus extends EventEmitter {
+  private requestsServer: WebSocket.Server;
+  private listenServer: WebSocket.Server;
+  private clientSockets: WebSocket[] = [];
+  private listeningSockets: WebSocket[] = [];
+
+  private unfulfilledRequests: Map<string, UnfulfilledRequest> = new Map([]);
+
+  constructor(
+    public requestsPort: number,
+    public listenPort: number,
+    public host: string = "0.0.0.0"
+  ) {
+    super();
+  }
+
+  async start() {
+    this.listenServer = await startWebSocketServer({
+      host: this.host,
+      port: this.listenPort
+    });
+
+    this.listenServer.on("connection", (newListener: WebSocket) => {
+      newListener.on("close", () => {
+        this.logToClients("Listener disconnected", "connections");
+
+        this.listeningSockets = this.listeningSockets.filter(
+          listener => listener !== newListener
+        );
+        this.terminateIfNoConnections();
+      });
+
+      // Require the listener to send a message *first* before being marked as ready
+      newListener.once("message", () => this.addNewListeneningSocket(newListener));
+    });
+
+    this.requestsServer = await startWebSocketServer({
+      host: this.host,
+      port: this.requestsPort
+    });
+
+    this.requestsServer.on("connection", (newClient: WebSocket) => {
+      newClient.on("message", (data: WebSocket.Data) => {
+        this.processRequest(newClient, data, this.listeningSockets);
+      });
+
+      newClient.on("close", () => {
+        this.logToClients("Client disconnected", "connections");
+
+        this.clientSockets = this.clientSockets.filter(
+          client => client !== newClient
+        );
+        this.clearClientRequests(newClient);
+        this.terminateIfNoConnections();
+      });
+
+      this.logToClients("Client connected", "connections");
+
+      this.clientSockets.push(newClient);
+    });
+  }
+
+  async ready() {
```

Comment:
Would we ever want this to time out if something is not working properly?

Reply:
For now we'll leave it like this.